core/src/main/java/io/smallrye/context/SingleWriterCopyOnWriteArrayIdentityMap.java

```diff
@@ -0,0 +1,101 @@
+package io.smallrye.context;
+
+import java.util.Arrays;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+/**
+ * This lookup map is optimized for the case where there is a single writer and multiple readers.
+ * <p>
+ * This map is not thread-safe for multiple writers and is not optimized for many entries, given
+ * that is backed by an ordered array and uses linear search.
+ */
+final class SingleWriterCopyOnWriteArrayIdentityMap<K, V> {
+
+    private static final AtomicReferenceFieldUpdater<SingleWriterCopyOnWriteArrayIdentityMap, Object[]> ENTRIES_UPDATER = AtomicReferenceFieldUpdater
+            .newUpdater(SingleWriterCopyOnWriteArrayIdentityMap.class, Object[].class, "entries");
+
+    private static final Object[] EMPTY_ARRAY = new Object[0];
+
+    private volatile Object[] entries;
+
+    public SingleWriterCopyOnWriteArrayIdentityMap() {
+        ENTRIES_UPDATER.lazySet(this, EMPTY_ARRAY);
+    }
+
+    public V get(K key) {
+        final Object[] array = this.entries;
+        for (int i = 0; i < array.length; i += 2) {
+            if (array[i] == key) {
+                return (V) array[i + 1];
+            }
+        }
+        return null;
+    }
+
+    public void put(K key, V value) {
+        Object[] oldEntries = entries;
+        // verify if the key already exists in the array
+        // or if the value is the same
+        int keyIndex = -1;
+        for (int i = 0; i < oldEntries.length; i += 2) {
+            if (oldEntries[i] == key) {
+                if (oldEntries[i + 1] == value) {
+                    return;
+                }
+                keyIndex = i;
+                break;
+            }
+        }
+        final Object[] newEntries;
+        if (keyIndex != -1) {
+            // key already exists, but the value is different
+            newEntries = new Object[oldEntries.length];
+            System.arraycopy(oldEntries, 0, newEntries, 0, oldEntries.length);
+            newEntries[keyIndex + 1] = value;
+        } else {
+            // key does not exist, add it
+            newEntries = new Object[oldEntries.length + 2];
+            System.arraycopy(oldEntries, 0, newEntries, 0, oldEntries.length);
+            newEntries[oldEntries.length] = key;
+            newEntries[oldEntries.length + 1] = value;
+        }
+        ENTRIES_UPDATER.lazySet(this, newEntries);
```

Comment:
Could you leave a comment here explaining why 'lazySet' specifically is strong enough for this use case?

Reply:
Sure thing, the lazyXYZ is the worst ever named concurrency method. And I pr try bad at name things eh...